scarb/src/ops/new.rs

```diff
@@ -0,0 +1,152 @@
+            name: name.clone(),
+        },
+        config,
+    )
+    .with_context(|| {
+        format!(
+            "failed to create package `{name}` at `{}`",
+            opts.path.display()
+        )
+    })?;
+
+    Ok(NewResult { name })
+}
+
+pub fn init_package(opts: InitOptions, config: &Config) -> Result<NewResult> {
+    ensure!(
+        !opts.path.join(MANIFEST_FILE_NAME).exists(),
+        "`scarb init` cannot be run on existing Scarb packages"
+    );
+
+    let name = infer_name(opts.name, &opts.path)?;
+    check_name(&name)?;
+
+    mk(
+        MkOpts {
+            path: opts.path,
+            name: name.clone(),
+        },
+        config,
+    )
+    .with_context(|| format!("failed to create package `{name}`",))?;
+
+    Ok(NewResult { name })
+}
+
+fn infer_name(name: Option<PackageName>, path: &Path) -> Result<PackageName> {
+    if let Some(name) = name {
+        Ok(name)
+    } else {
+        let Some(file_name) = path.file_name() else {
+            bail!(
+                "cannot infer package name from path {:?}\nUse --name to override.",
+                path.as_os_str()
+            );
+        };
+
+        let Some(file_name) = file_name.to_str() else {
+            bail!("cannot create package with a non-unicode name: {:?}", file_name);
+        };
+
+        Ok(file_name.into())
+    }
+}
+
+fn check_name(name: &PackageName) -> Result<()> {
+    restricted_names::validate_package_name(&name, "package name")?;
+    // TODO(mkaput): Check not Cairo keyword.
+    // TODO(mkaput): Warn if name is `corelib`.
+    Ok(())
+}
+
+struct MkOpts {
+    path: PathBuf,
+    name: PackageName,
+}
+
+fn mk(MkOpts { path, name }: MkOpts, _config: &Config) -> Result<()> {
+    // Create project directory in case we are called from `new` op.
+    fsx::create_dir_all(&path)?;
+
+    // Write VCS ignore file.
+    fsx::write(path.join(".gitignore"), DEFAULT_TARGET_DIR_NAME)?;
+
+    // Create the `Scarb.toml` file.
+    fsx::write(
+        path.join(MANIFEST_FILE_NAME),
+        formatdoc! {r#"
+            [package]
+            name = "{name}"
+            version = "0.1.0"
+
+            # See more keys and their definitions at https://github.com/software-mansion/scarb/blob/main/scarb/src/core/manifest/toml.rs
+
+            [dependencies]
+            # foo = {{ path = "vendor/foo" }}
+        "#},
+    )?;
+
+    // Create hello world source files (with respective parent directories) if source directory
+    // does not exist.
+    let source_dir = path.join(DEFAULT_SOURCE_DIR_NAME);
+    if !source_dir.exists() {
+        fsx::create_dir_all(&source_dir)?;
+
+        fsx::write(
+            source_dir.join("lib.cairo"),
+            indoc! {r#"
+                fn fib(a: felt, b: felt, n: felt) -> felt {
+                    match n {
+                        0 => a,
+                        _ => fib(b, a + b, n - 1),
+                    }
+                }
+            "#},
+        )?;
+    }
```

Comment:
On a similar note, should we overwrite source files that may be created by user before using `init`? I presume one may want to use `init` on a project that has already been populated with some source files.

Reply:
There's `    if !source_dir.exists() {` to check this

Reply:
https://github.com/software-mansion/scarb/pull/36/commits/2a56e8479d933c8a657ec7a9fb6cb500db9ee31e